File: UISamples/Bit6UIFullDemo-ObjC/ChatDemo/InCall/CustomCallViewController.m
//
//  BXUCallViewController.m
//  Bit6
//
//  Created by Carlos Thurber on 01/05/15.
//  Copyright (c) 2015 Bit6. All rights reserved.
//

#import "CustomCallViewController.h"
#import <AVFoundation/AVFoundation.h>

@interface CustomCallViewController ()

@property (weak, nonatomic) UILabel *noLocalFeedLabel;

@property (weak, nonatomic) IBOutlet UIView *overlayView;
@property (weak, nonatomic) IBOutlet UIView *controlsView;

@property (weak, nonatomic) IBOutlet BXUContactNameLabel *usernameLabel;
@property (weak, nonatomic) IBOutlet UILabel *timerLabel;

@property (weak, nonatomic) IBOutlet BXUCircleButton *muteAudioButton;
@property (weak, nonatomic) IBOutlet BXUCircleButton *muteVideoButton;
@property (weak, nonatomic) IBOutlet BXUCircleButton *bluetoothButton;
@property (weak, nonatomic) IBOutlet BXUCircleButton *speakerButton;
@property (weak, nonatomic) IBOutlet BXUCircleButton *cameraButton;
@property (weak, nonatomic) IBOutlet BXUCircleButton *recordingCallButton;

@end

@implementation CustomCallViewController

+ (Bit6CallViewController*)createForCall:(Bit6CallController*)callController
{
    return [[BXUCallViewController alloc] initWithNibName:@"CustomCallViewController" bundle:nil];
}

- (Bit6CallController*) callController
{
    return [[Bit6 activeCalls] firstObject];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self refreshState];
    
    if ([Bit6 activeCalls].count>1) {
        self.usernameLabel.address = nil;
        self.usernameLabel.text = @"Many Destinations";
    }
    else {
        self.usernameLabel.address = self.callController.other;
    }
    
    [self.bluetoothButton setTitle:@"" forState:UIControlStateNormal];
    [self.bluetoothButton setBackgroundImage:[UIImage imageNamed:@"Bit6UIResources.bundle/bluetooth"] forState:UIControlStateNormal];
    [self.speakerButton setTitle:@"" forState:UIControlStateNormal];
    [self.speakerButton setBackgroundImage:[UIImage imageNamed:@"Bit6UIResources.bundle/speaker"] forState:UIControlStateNormal];
    [self.muteAudioButton setTitle:@"" forState:UIControlStateNormal];
    [self.muteAudioButton setBackgroundImage:[UIImage imageNamed:@"Bit6UIResources.bundle/mute"] forState:UIControlStateNormal];
    [self.muteVideoButton setTitle:@"" forState:UIControlStateNormal];
    [self.muteVideoButton setBackgroundImage:[UIImage imageNamed:@"Bit6UIResources.bundle/mute_video"] forState:UIControlStateNormal];
    [self.cameraButton setTitle:@"" forState:UIControlStateNormal];
    [self.cameraButton setBackgroundImage:[UIImage imageNamed:@"Bit6UIResources.bundle/camera"] forState:UIControlStateNormal];
    [self.recordingCallButton setTitle:@"" forState:UIControlStateNormal];
    [self.recordingCallButton setBackgroundImage:[UIImage imageNamed:@"Bit6UIResources.bundle/record_call"] forState:UIControlStateNormal];
    
    UITapGestureRecognizer *tgr = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(overlayTapped:)];
    [self.overlayView addGestureRecognizer:tgr];
    UITapGestureRecognizer *tgr2 = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(overlayTapped:)];
    [self.view addGestureRecognizer:tgr2];
    
    BOOL atLeastOneCallConnected = NO;
    NSArray<Bit6CallController*>* callControllers = [Bit6 activeCalls];
    for (Bit6CallController *call in callControllers) {
        if (call.state >= Bit6CallState_CONNECTED) {
            atLeastOneCallConnected = YES; break;
        }
    }
    self.controlsView.hidden = !atLeastOneCallConnected;
}

- (void)overlayTapped:(UITapGestureRecognizer*)tgr
{
    BOOL atLeastOneCallConnected = NO;
    BOOL atLeastOneCallHasVideo = NO;
    NSArray<Bit6CallController*>* callControllers = [Bit6 activeCalls];
    for (Bit6CallController *call in callControllers) {
        if (call.state >= Bit6CallState_CONNECTED) {
            atLeastOneCallConnected = YES;
        }
        if (call.hasVideo) {
            atLeastOneCallHasVideo = YES;
        }
    }
    
    if (atLeastOneCallHasVideo) {
        if (!self.overlayView.hidden) {
            if (atLeastOneCallConnected) {
                self.overlayView.hidden = !self.overlayView.hidden;
            }
        }
        else {
            self.overlayView.hidden = !self.overlayView.hidden;
        }
    }
}

#pragma mark - Bit6CallControllerDelegate

- (void)callController:(Bit6CallController*)callController callDidChangeToState:(Bit6CallState)state
{
    [super callController:callController callDidChangeToState:state];
    
    if (callController.error) {
        [Bit6AlertView showAlertControllerWithTitle:callController.error.localizedDescription message:nil cancelButtonTitle:@"OK"];
    }
    
    [self refreshState];
    
    [self secondsDidChangeForCallController:callController];
}

- (void)secondsDidChangeForCallController:(Bit6CallController*)callController
{
    [super secondsDidChangeForCallController:callController];
    
    CGFloat longerCall = 0.0f;
    NSArray<Bit6CallController*>* callControllers = [Bit6 activeCalls];
    for (Bit6CallController *call in callControllers) {
        if (call.seconds > longerCall) {
            longerCall = call.seconds;
        }
    }
    
    if (self.callController.state == Bit6CallState_CONNECTED) {
        self.timerLabel.text = [Bit6Utils clockFormatForSeconds:longerCall];
    }
}

- (void)callController:(Bit6CallController*)callController localVideoFeedInterruptedBecause:(int)reason
{
    [super callController:callController localVideoFeedInterruptedBecause:reason];
    
    if (!self.noLocalFeedLabel) {
        UILabel *label = [[UILabel alloc] initWithFrame:CGRectZero];
        label.backgroundColor = [UIColor grayColor];
        label.tag = 15;
        label.numberOfLines = 2;
        label.text = @"Video Feed\nUnavailable";
        label.font = [UIFont fontWithName:@"HelveticaNeue-Medium" size:16];
        label.autoresizingMask = UIViewAutoresizingFlexibleWidth|UIViewAutoresizingFlexibleHeight|UIViewAutoresizingFlexibleTopMargin|UIViewAutoresizingFlexibleBottomMargin|UIViewAutoresizingFlexibleLeftMargin|UIViewAutoresizingFlexibleRightMargin;
        label.textAlignment = NSTextAlignmentCenter;
        label.adjustsFontSizeToFitWidth = YES;
        label.minimumScaleFactor = 0.5;
        label.textColor = [UIColor whiteColor];
        
        [self.localVideoView addSubview:label];
        label.frame = self.localVideoView.bounds;
        self.noLocalFeedLabel = label;
    }
}

- (void)localVideoFeedInterruptionEndedForCallController:(Bit6CallController*)callController
{
    [super localVideoFeedInterruptionEndedForCallController:callController];
    
    [self.noLocalFeedLabel removeFromSuperview];
}

#pragma mark - Bit6CallViewController methods

- (void)refreshState
{
    Bit6CallController *smallerCall = nil;
    Bit6CallState smallerState = Bit6CallState_END;
    NSArray<Bit6CallController*>* callControllers = [Bit6 activeCalls];
    for (Bit6CallController *call in callControllers) {
        if (call.state < smallerState) {
            smallerState = call.state;
            smallerCall = call;
        }
        if (call.state == Bit6CallState_CONNECTED) {
            self.controlsView.hidden = NO;
        }
    }
    
    if (smallerCall.incoming) {
        switch (smallerState) {
            case Bit6CallState_NEW:
            case Bit6CallState_ACCEPTING_CALL:
                self.timerLabel.text = @"Answering Call..."; break;
            case Bit6CallState_WAITING_SDP:
                self.timerLabel.text = @"Waiting SDP..."; break;
            case Bit6CallState_GATHERING_CANDIDATES:
                self.timerLabel.text = @"Gathering Candidates..."; break;
            case Bit6CallState_SENDING_SDP:
                self.timerLabel.text = @"Sending SDP..."; break;
            case Bit6CallState_CONNECTING:
                self.timerLabel.text = @"Connecting..."; break;
            case Bit6CallState_CONNECTED: break;
            case Bit6CallState_DISCONNECTED:
                self.timerLabel.text = @"Disconnected"; break;
            case Bit6CallState_END:
            default: break;
        }
    }
    else {
        switch (smallerState) {
            case Bit6CallState_NEW:
            case Bit6CallState_GATHERING_CANDIDATES:
                self.timerLabel.text = @"Gathering Candidates..."; break;
            case Bit6CallState_SENDING_SDP:
                self.timerLabel.text = @"Sending SDP..."; break;
            case Bit6CallState_WAITING_SDP:
                self.timerLabel.text = @"Waiting for Answer..."; break;
            case Bit6CallState_CONNECTING:
                self.timerLabel.text = @"Connecting..."; break;
            case Bit6CallState_CONNECTED: break;
            case Bit6CallState_DISCONNECTED:
                self.timerLabel.text = @"Disconnected"; break;
            case Bit6CallState_END:
            default: break;
        }
    }
}

- (void)refreshControlsView
{
    BOOL atLeastOneCallHasVideo = NO;
    BOOL atLeastOneCallHasAudio = NO;
    BOOL atLeastOneCallHasRemoteAudio = NO;
    NSArray<Bit6CallController*>* callControllers = [Bit6 activeCalls];
    for (Bit6CallController *call in callControllers) {
        if (call.hasVideo) {
            atLeastOneCallHasVideo = YES;
        }
        if (call.hasAudio) {
            atLeastOneCallHasAudio = YES;
        }
        if (call.hasRemoteAudio) {
            atLeastOneCallHasRemoteAudio = YES;
        }
    }
    
    if (TARGET_OS_SIMULATOR) {
        self.muteVideoButton.enabled = NO;
        self.cameraButton.enabled = NO;
        self.bluetoothButton.enabled = NO;
        self.speakerButton.enabled = NO;
    }
    else {
        self.muteVideoButton.enabled = self.localVideoView.interrupted ? NO : atLeastOneCallHasVideo;
        self.cameraButton.enabled = atLeastOneCallHasVideo;
        self.bluetoothButton.enabled = atLeastOneCallHasRemoteAudio;
        self.speakerButton.enabled = atLeastOneCallHasRemoteAudio;
    }
    
    self.recordingCallButton.enabled = callControllers.count==1 && [self.callController supportsCapability:Bit6CallCapability_Recording];
    
    self.muteAudioButton.enabled = atLeastOneCallHasAudio;
    
    NSString *deviceType = [UIDevice currentDevice].model;
    if(![deviceType isEqualToString:@"iPhone"]){
        self.speakerButton.enabled = NO;
    }
    
    if (self.muteAudioButton.enabled) {
        self.muteAudioButton.on = ![Bit6CallController isLocalAudioEnabled];
    }
    if (self.bluetoothButton.enabled) {
        self.bluetoothButton.on = [Bit6CallController isBluetoothEnabled];
    }
    if (self.speakerButton.enabled) {
        self.speakerButton.on = [Bit6CallController isSpeakerEnabled];
    }
    if (self.muteVideoButton.enabled) {
        self.muteVideoButton.on = ![Bit6CallController isLocalVideoEnabled];
    }
    if (self.recordingCallButton.enabled) {
        self.recordingCallButton.on = self.callController.recording;
    }
}

- (void)updateLayoutForVideoFeedViews:(NSArray*)videoFeedViews
{
    [super updateLayoutForVideoFeedViews:videoFeedViews];
    
    if ([Bit6 activeCalls].count>1) {
        self.usernameLabel.address = nil;
        self.usernameLabel.text = @"Many Destinations";
    }
    else {
        self.usernameLabel.address = self.callController.other;
    }
    
    self.noLocalFeedLabel.frame = self.localVideoView.bounds;
}

#pragma mark Actions

- (IBAction)muteAudioCall:(id)sender {
    [Bit6CallController setLocalAudioEnabled:!Bit6CallController.isLocalAudioEnabled];
}

- (IBAction)bluetooth:(id)sender {
    [Bit6CallController setBluetoothEnabled:!Bit6CallController.isBluetoothEnabled];
}

- (IBAction)speaker:(id)sender {
    [Bit6CallController setSpeakerEnabled:!Bit6CallController.isSpeakerEnabled];
}

- (IBAction)muteVideoCall:(id)sender {
    [Bit6CallController setLocalVideoEnabled:!Bit6CallController.isLocalVideoEnabled];
}

- (IBAction)switchCamera:(id)sender {
    [Bit6CallController setLocalVideoSource:[Bit6CallController localVideoSource]==Bit6VideoSource_CameraBack ? Bit6VideoSource_CameraFront : Bit6VideoSource_CameraBack];
}

- (IBAction)switchRecording:(id)sender {
    self.callController.recording = !self.callController.recording;
}

- (IBAction)hangup:(id)sender {
    [Bit6CallController hangupAll];
}

@end

File: UISamples/Bit6UIFullDemo-Swift/Bit6ChatDemo-Swift/CustomCallViewController.swift
//
//  CustomCallViewController.swift
//  Bit6UIFullDemo
//
//  Created by Carlos Thurber on 08/25/16.
//  Copyright © 2016 Bit6. All rights reserved.
//

import UIKit
import Bit6
import Bit6UI

class CustomCallViewController: Bit6CallViewController {
    
    var noLocalFeedLabel : UILabel?
    
    @IBOutlet var overlayView:UIView!
    @IBOutlet var controlsView:UIView!
    
    @IBOutlet var usernameLabel:BXUContactNameLabel!
    @IBOutlet var timerLabel:UILabel!
    
    @IBOutlet var muteAudioButton:BXUCircleButton!
    @IBOutlet var bluetoothButton:BXUCircleButton!
    @IBOutlet var speakerButton:BXUCircleButton!
    @IBOutlet var muteVideoButton:BXUCircleButton!
    @IBOutlet var cameraButton:BXUCircleButton!
    @IBOutlet var recordingCallButton:BXUCircleButton!
    
    override class func createForCall(callController:Bit6CallController) -> Bit6CallViewController {
        return CustomCallViewController(nibName:"CustomCallViewController", bundle:nil)
    }
    
    var callController : Bit6CallController? {
        get {
            return self.callControllers.first
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        refreshState()
        
        if self.callControllers.count>1 {
            self.usernameLabel.address = nil
            self.usernameLabel.text = "Many Destinations"
        }
        else {
            self.usernameLabel.address = self.callController?.other
        }
        
        bluetoothButton.setTitle("", forState:.Normal)
        bluetoothButton.setBackgroundImage(UIImage(named:"Bit6UIResources.bundle/bluetooth"), forState:.Normal)
        speakerButton.setTitle("", forState:.Normal)
        speakerButton.setBackgroundImage(UIImage(named:"Bit6UIResources.bundle/speaker"), forState:.Normal)
        muteAudioButton.setTitle("", forState:.Normal)
        muteAudioButton.setBackgroundImage(UIImage(named:"Bit6UIResources.bundle/mute"), forState:.Normal)
        muteVideoButton.setTitle("", forState:.Normal)
        muteVideoButton.setBackgroundImage(UIImage(named:"Bit6UIResources.bundle/mute_video"), forState:.Normal)
        cameraButton.setTitle("", forState:.Normal)
        cameraButton.setBackgroundImage(UIImage(named:"Bit6UIResources.bundle/camera"), forState:.Normal)
        recordingCallButton.setTitle("", forState:.Normal)
        recordingCallButton.setBackgroundImage(UIImage(named:"Bit6UIResources.bundle/record_call"), forState:.Normal)
        
        let tgr = UITapGestureRecognizer(target:self, action:#selector(CustomCallViewController.overlayTapped(_:)))
        overlayView.addGestureRecognizer(tgr)
        
        let tgr2 = UITapGestureRecognizer(target:self, action:#selector(CustomCallViewController.overlayTapped(_:)))
        view.addGestureRecognizer(tgr2)
        
        var atLeastOneCallConnected = false
        let callControllers = self.callControllers
        for call in callControllers {
            if call.state.rawValue >= Bit6CallState.CONNECTED.rawValue {
                atLeastOneCallConnected = true
                break
            }
        }
        
        controlsView.hidden = !atLeastOneCallConnected
    }
    
    func overlayTapped(tgr:UITapGestureRecognizer){
        var atLeastOneCallConnected = false
        var atLeastOneCallHasVideo = false
        
        let callControllers = self.callControllers
        for call in callControllers {
            if call.state.rawValue >= Bit6CallState.CONNECTED.rawValue {
                atLeastOneCallConnected = true
            }
            if call.hasVideo {
                atLeastOneCallHasVideo = true
            }
        }
        
        if (atLeastOneCallHasVideo) {
            if (!overlayView.hidden) {
                if (atLeastOneCallConnected) {
                    overlayView.hidden = !overlayView.hidden
                }
            }
            else {
                overlayView.hidden = !overlayView.hidden
            }
        }
    }
    
    // MARK: Bit6CallControllerDelegate
    
    override func callController(callController: Bit6CallController, callDidChangeToState state: Bit6CallState) {
        super.callController(callController, callDidChangeToState:state)
        
        if let error = callController.error {
            Bit6AlertView.showAlertControllerWithTitle(error.localizedDescription, message:nil, cancelButtonTitle:"OK")
        }
        
        if isViewLoaded() {
            refreshState()
        }
        
        secondsDidChangeForCallController(callController)
    }
    
    override func secondsDidChangeForCallController(callController: Bit6CallController) {
        super.secondsDidChangeForCallController(callController)
        
        if isViewLoaded() {
            var longerCall : UInt = 0
            let callControllers = self.callControllers
            for call in callControllers {
                if call.seconds > longerCall {
                    longerCall = call.seconds
                }
            }
            
            if callController.state == .CONNECTED {
                timerLabel.text = Bit6Utils.clockFormatForSeconds(Double(longerCall))
            }
        }
    }
    
    override func callController(callController: Bit6CallController, localVideoFeedInterruptedBecause reason: Int32) {
        super.callController(callController, localVideoFeedInterruptedBecause: reason)
        
        if self.noLocalFeedLabel == nil {
            if let localVideoView = self.localVideoView {
                let label = UILabel(frame:CGRectZero)
                label.backgroundColor = UIColor.grayColor()
                label.tag = 15
                label.numberOfLines = 2
                label.text = "Video Feed\nUnavailable"
                label.font = UIFont(name:"HelveticaNeue-Medium", size:16)
                label.autoresizingMask = [.FlexibleWidth,.FlexibleHeight,.FlexibleTopMargin,.FlexibleBottomMargin,.FlexibleLeftMargin,.FlexibleRightMargin]
                label.textAlignment = .Center
                label.adjustsFontSizeToFitWidth = true
                label.minimumScaleFactor = 0.5
                label.textColor = UIColor.whiteColor()
                
                localVideoView.addSubview(label)
                label.frame = localVideoView.bounds
                self.noLocalFeedLabel = label
            }
        }
    }
    
    override func localVideoFeedInterruptionEndedForCallController(callController: Bit6CallController) {
        super.localVideoFeedInterruptionEndedForCallController(callController)
        
        self.noLocalFeedLabel?.removeFromSuperview()
    }
    
    // MARK: Bit6CallViewController methods
    
    func refreshState(){
        var smallerCall : Bit6CallController? = nil
        var smallerState = Bit6CallState.END
        
        let callControllers = self.callControllers
        for call in callControllers {
            if call.state.rawValue < smallerState.rawValue {
                smallerState = call.state
                smallerCall = call
            }
            if call.state == Bit6CallState.CONNECTED {
                controlsView.hidden = false
            }
        }
        
        if let smallerCall = smallerCall {
            if smallerCall.incoming {
                switch smallerState {
                case .NEW: fallthrough
                case .ACCEPTING_CALL:
                    timerLabel.text = "Answering Call..."
                case .WAITING_SDP:
                    timerLabel.text = "Waiting SDP..."
                case .GATHERING_CANDIDATES:
                    timerLabel.text = "Gathering Candidates..."
                case .SENDING_SDP:
                    timerLabel.text = "Sending SDP..."
                case .CONNECTING:
                    timerLabel.text = "Connecting..."
                case .CONNECTED: break
                case .DISCONNECTED:
                    timerLabel.text = "Disconnected"
                case .END: fallthrough
                default : break
                }
            }
            else {
                switch (smallerState) {
                case .NEW: fallthrough
                case .GATHERING_CANDIDATES:
                    timerLabel.text = "Gathering Candidates..."
                case .SENDING_SDP:
                    timerLabel.text = "Sending SDP..."
                case .WAITING_SDP:
                    timerLabel.text = "Waiting for Answer..."
                case .CONNECTING:
                    timerLabel.text = "Connecting..."
                case .CONNECTED: break;
                case .DISCONNECTED:
                    timerLabel.text = "Disconnected"
                case .END: fallthrough
                default: break;
                }
            }
        }
    }
    
    override func refreshControlsView() {
        var atLeastOneCallHasVideo = false
        var atLeastOneCallHasAudio = false
        var atLeastOneCallHasRemoteAudio = false
        
        let callControllers = self.callControllers
        for call in callControllers {
            if call.hasVideo {
                atLeastOneCallHasVideo = true
            }
            if call.hasAudio {
                atLeastOneCallHasAudio = true
            }
            if call.hasRemoteAudio {
                atLeastOneCallHasRemoteAudio = true
            }
        }
        
        if (TARGET_OS_SIMULATOR != 0) {
            self.muteVideoButton.enabled = false
            self.cameraButton.enabled = false
            self.bluetoothButton.enabled = false
            self.speakerButton.enabled = false
        }
        else {
            self.muteVideoButton.enabled = (self.localVideoView?.interrupted ?? false) ? false : atLeastOneCallHasVideo
            self.cameraButton.enabled = atLeastOneCallHasVideo
            self.bluetoothButton.enabled = atLeastOneCallHasRemoteAudio
            self.speakerButton.enabled = atLeastOneCallHasRemoteAudio
        }
        
        self.recordingCallButton?.enabled = callControllers.count==1 && self.callController!.supportsCapability(.Recording)
        
        muteAudioButton.enabled = atLeastOneCallHasAudio
        
        let deviceType = UIDevice.currentDevice().model
        if deviceType != "iPhone" {
            speakerButton.enabled = false
        }
        
        if muteAudioButton.enabled {
            muteAudioButton.on = !Bit6CallController.isLocalAudioEnabled()
        }
        if bluetoothButton.enabled {
            bluetoothButton.on = Bit6CallController.isBluetoothEnabled()
        }
        if speakerButton.enabled {
            speakerButton.on = Bit6CallController.isSpeakerEnabled()
        }
        if muteVideoButton.enabled {
            muteVideoButton.on = Bit6CallController.isLocalVideoEnabled()
        }
        if recordingCallButton.enabled {
            recordingCallButton.on = self.callController!.recording
        }
    }
    
    override func updateLayoutForVideoFeedViews(videoFeedViews: [Bit6VideoFeedView]) {
        super.updateLayoutForVideoFeedViews(videoFeedViews)
        
        if self.callControllers.count>1 {
            self.usernameLabel.address = nil
            self.usernameLabel.text = "Many Destinations"
        }
        else {
            self.usernameLabel.address = self.callController?.other
        }
        
        if let localViewView = self.localVideoView {
            self.noLocalFeedLabel?.frame = localViewView.bounds
        }
    }
    
    // MARK: Actions
    
    @IBAction func muteAudioCall(sender : UIButton) {
        Bit6CallController.setLocalAudioEnabled(!Bit6CallController.isLocalAudioEnabled())
    }
    
    @IBAction func bluetooth(sender : UIButton) {
        Bit6CallController.setBluetoothEnabled(!Bit6CallController.isBluetoothEnabled())
    }
    
    @IBAction func speaker(sender : UIButton) {
        Bit6CallController.setSpeakerEnabled(!Bit6CallController.isSpeakerEnabled())
    }
    
    @IBAction func muteVideoCall(sender : UIButton) {
        Bit6CallController.setLocalVideoEnabled(!Bit6CallController.isLocalVideoEnabled())
    }
    
    @IBAction func switchCamera(sender : UIButton) {
        Bit6CallController.setLocalVideoSource(Bit6CallController.localVideoSource() == .CameraBack ? .CameraFront : .CameraBack)
    }
    
    @IBAction func switchRecording(sender : UIButton) {
        self.callController!.recording = !self.callController!.recording
    }
    
    @IBAction func hangup(sender : UIButton) {
        Bit6CallController.hangupAll()
    }
    
}
